include/swift/AST/SourceFile.h

```diff
@@ -117,6 +117,13 @@ class SourceFile final : public FileUnit {
   /// The scope map that describes this source file.
   std::unique_ptr<ASTScope> Scope;
 
+  /// The set of validated opaque return type decls in the source file.
+  llvm::SmallVector<OpaqueTypeDecl *, 4> OpaqueReturnTypes;
```

Comment:
I think we can drop this if the request were kicked off on lookup of the individual value decl.  This would also let you drop the validated/unvalidated distinction.

Reply:
This is actually used by IRGen when emitting all the opaque types of a source file. Otherwise, it would need to walk the SourceFile recursively again to find opaque types inside local nested types, etc.